middleware/http_tracing.go

```diff
@@ -50,3 +46,111 @@ func (t Tracer) Wrap(next http.Handler) http.Handler {
+// wrapped into gRPC requests and routed through the httpgrpc.HTTP/Handle gRPC.
+// HTTPGRPCTracer.Wrap does not need to be attached to server.Config.HTTPMiddleware, which
+// uses Tracer.Wrap for the standard case of direct HTTP requests not routed through gRPC.
+func InitHTTPGRPCMiddleware(router *mux.Router) *mux.Router {
+	middleware := HTTPGRPCTracer{RouteMatcher: router}
+	router.Use(middleware.Wrap)
+	return router
+}
+
+// Wrap creates and decorates server-side tracing spans for httpgrpc requests
+//
+// The httpgrpc client wraps HTTP requests up into a generic httpgrpc.HTTP/Handle gRPC method.
+// The httpgrpc server unwraps httpgrpc.HTTP/Handle gRPC requests into HTTP requests
+// and forwards them to its own internal HTTP router.
+//
+// By default, the server-side tracing spans for the httpgrpc.HTTP/Handle gRPC method
+// have no data about the wrapped HTTP request being handled.
+//
+// HTTPGRPCTracer.Wrap starts a child span with span name and tags following the approach in
+// Tracer.Wrap's usage of opentracing-contrib/go-stdlib/nethttp.Middleware
+// and attaches the HTTP server span tags to the parent httpgrpc.HTTP/Handle gRPC span, allowing
+// tracing tooling to differentiate the HTTP requests represented by the httpgrpc.HTTP/Handle spans.
+//
+// opentracing-contrib/go-stdlib/nethttp.Middleware could not be used here
+// as it does not expose options to access and tag the incoming parent span.
+//
+// Parent span tagging depends on using a Jaeger tracer for now to check the parent span's
+// OperationName(), which is not available on the generic opentracing Tracer interface.
+func (hgt HTTPGRPCTracer) Wrap(next http.Handler) http.Handler {
+	fn := func(w http.ResponseWriter, r *http.Request) {
+		ctx := r.Context()
+		tracer := opentracing.GlobalTracer()
+
+		// skip spans which were not forwarded from non-httpgrpc.HTTP/Handle spans;
+		// standard http spans started directly from the HTTP server are presumed to
+		// already be instrumented by Tracer.Wrap
+		parentSpan := opentracing.SpanFromContext(ctx)
+		if parentSpan, ok := parentSpan.(*jaeger.Span); ok {
+			if parentSpan.OperationName() != httpGRPCHandleMethod {
+				next.ServeHTTP(w, r)
+				return
+			}
+		}
+
+		// extract relevant span & tag data from request
+		method := r.Method
+		matchedRoute := getRouteName(hgt.RouteMatcher, r)
+		url := r.URL.String()
+		userAgent := r.Header.Get("User-Agent")
+
+		// tag parent httpgrpc.HTTP/Handle server span, if it exists
+		if parentSpan != nil {
+			parentSpan.SetTag(string(ext.HTTPUrl), url)
+			parentSpan.SetTag(string(ext.HTTPMethod), method)
+			parentSpan.SetTag("http.route", matchedRoute)
+			parentSpan.SetTag("http.user_agent", userAgent)
+		}
+
+		// create and start child HTTP span
+		// mirroring opentracing-contrib/go-stdlib/nethttp.Middleware span name and tags
+		childSpanName := makeHTTPOperationNameFunc(hgt.RouteMatcher)(r)
```

Comment:
Should we store the result of `makeHTTPOperationNameFunc` on `HTTPGRPCTracer` once (eg. in `InitHTTPGRPCMiddleware`)? Then we could do something like this:

```suggestion
		childSpanName := hgt.makeHTTPOperationName(r)
```

Reply:
I would lean towards *not* exposing the span naming scheme that as something a user could change, because then the two opentracing middlewares that are both active could be instantiated with different patterns of forming the operation name.

As long as the span naming is standards compliant, I think it's alright for the library to make that decision for the user, or at least to not allow the user to make two different decisions for two middlewares that are supposed to behave identically.

Reply:
Sorry, I should have been clearer in my comment: I don't think we should expose this to the consumer. I'm wondering if we can just call `makeHTTPOperationNameFunc` once in `InitHTTPGRPCMiddleware`, rather than calling it every time in this method.

Reply:
just updated - I inlined it in the outer `Wrap` method, rather than the returned middleware method, so it doesn't get called for every request. It now mirrors how the standard HTTP tracer above it works.

`InitHTTPGRPCMiddleware` is just a convenience func that doesn't necessarily need to be used to apply the middleware - putting a call to a private `makeHTTPOperationNameFunc` in there would make `InitHTTPGRPCMiddleware` instead of optional.

Reply:
Even better, thanks for addressing this.